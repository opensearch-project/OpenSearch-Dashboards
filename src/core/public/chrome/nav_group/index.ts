/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export {
  ChromeNavGroupService,
  ChromeNavGroupServiceSetupContract,
  ChromeNavGroupServiceStartContract,
  ChromeRegistrationNavLink,
  NavGroupItemInMap,
  ChromeNavGroupUpdater,
} from './nav_group_service';
