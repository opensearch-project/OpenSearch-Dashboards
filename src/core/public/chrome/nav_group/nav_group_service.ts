/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { BehaviorSubject, combineLatest, Observable, of, ReplaySubject, Subscription } from 'rxjs';
import { AppCategory, ChromeNavGroup, ChromeNavLink } from 'opensearch-dashboards/public';
import { map, switchMap, takeUntil } from 'rxjs/operators';
import { IUiSettingsClient } from '../../ui_settings';
import {
  flattenLinksOrCategories,
  fulfillRegistrationLinksToChromeNavLinks,
  getOrderedLinksOrCategories,
} from '../utils';
import { ChromeNavLinks } from '../nav_links';

/** @public */
export interface ChromeRegistrationNavLink {
  id: string;
  title?: string;
  category?: AppCategory;
  order?: number;

  /**
   * link with parentNavLinkId field will be displayed as nested items in navigation.
   */
  parentNavLinkId?: string;
}

export type NavGroupItemInMap = ChromeNavGroup & {
  navLinks: ChromeRegistrationNavLink[];
};

export type ChromeNavGroupUpdater = (navGroup: ChromeNavGroup) => Partial<ChromeNavGroup> | void;

export interface ChromeNavGroupServiceSetupContract {
  addNavLinksToGroup: (navGroup: ChromeNavGroup, navLinks: ChromeRegistrationNavLink[]) => void;
  /**
   * Get a boolean value to indicates whether use case is enabled
   */
  getNavGroupEnabled: () => boolean;
  registerNavGroupUpdater: (navGroupUpdater: Observable<ChromeNavGroupUpdater>) => () => void;
}

export interface ChromeNavGroupServiceStartContract {
  getNavGroupsMap$: () => Observable<Record<string, NavGroupItemInMap>>;
  getNavGroupEnabled: ChromeNavGroupServiceSetupContract['getNavGroupEnabled'];
}

/** @internal */
export class ChromeNavGroupService {
  private readonly navGroupsMap$ = new BehaviorSubject<Record<string, NavGroupItemInMap>>({});
  private readonly stop$ = new ReplaySubject(1);
  private navLinks$: Observable<Array<Readonly<ChromeNavLink>>> = new BehaviorSubject([]);
  private navGroupEnabled: boolean = false;
  private navGroupEnabledUiSettingsSubscription: Subscription | undefined;
  private navGroupUpdaters$$ = new BehaviorSubject<Array<Observable<ChromeNavGroupUpdater>>>([]);
  private addNavLinkToGroup(
    currentGroupsMap: Record<string, NavGroupItemInMap>,
    navGroup: ChromeNavGroup,
    navLink: ChromeRegistrationNavLink
  ) {
    const matchedGroup = currentGroupsMap[navGroup.id];
    if (matchedGroup) {
      const links = matchedGroup.navLinks;
      const isLinkExistInGroup = links.some((link) => link.id === navLink.id);
      if (isLinkExistInGroup) {
        // eslint-disable-next-line no-console
        console.warn(
          `[ChromeService] Navlink of ${navLink.id} has already been registered in group ${navGroup.id}`
        );
        return currentGroupsMap;
      }
      matchedGroup.navLinks.push(navLink);
    } else {
      currentGroupsMap[navGroup.id] = {
        ...navGroup,
        navLinks: [navLink],
      };
    }

    return currentGroupsMap;
  }
  private getSortedNavGroupsMap$() {
    return combineLatest([this.getUpdatedNavGroupsMap$(), this.navLinks$])
      .pipe(takeUntil(this.stop$))
      .pipe(
        map(([navGroupsMap, navLinks]) => {
          return Object.keys(navGroupsMap).reduce((sortedNavGroupsMap, navGroupId) => {
            const navGroup = navGroupsMap[navGroupId];
            const sortedNavLinks = getOrderedLinksOrCategories(
              fulfillRegistrationLinksToChromeNavLinks(navGroup.navLinks, navLinks)
            );
            sortedNavGroupsMap[navGroupId] = {
              ...navGroup,
              navLinks: flattenLinksOrCategories(sortedNavLinks),
            };
            return sortedNavGroupsMap;
          }, {} as Record<string, NavGroupItemInMap>);
        })
      );
  }

  private getUpdatedNavGroupsMap$() {
    return combineLatest([this.navGroupsMap$, this.navGroupUpdaters$$]).pipe(
      switchMap(([navGroupsMap, updaters$]) => {
        if (updaters$.length === 0) {
          return of(navGroupsMap);
        }
        return combineLatest(updaters$).pipe(
          map((updaters) => {
            return Object.keys(navGroupsMap).reduce<Record<string, NavGroupItemInMap>>(
              (previousValue, currentKey) => ({
                ...previousValue,
                [currentKey]: updaters.reduce(
                  (prevNavGroup, currentUpdater) => ({
                    ...prevNavGroup,
                    ...currentUpdater(prevNavGroup),
                  }),
                  navGroupsMap[currentKey]
                ),
              }),
              {}
            );
          })
        );
      })
    );
  }
  setup({ uiSettings }: { uiSettings: IUiSettingsClient }): ChromeNavGroupServiceSetupContract {
    this.navGroupEnabledUiSettingsSubscription = uiSettings
      .get$('home:useNewHomePage', false)
      .subscribe((value) => {
        this.navGroupEnabled = value;
      });

    return {
      addNavLinksToGroup: (navGroup: ChromeNavGroup, navLinks: ChromeRegistrationNavLink[]) => {
        // Construct a new groups map pointer.
        const currentGroupsMap = { ...this.navGroupsMap$.getValue() };

        const navGroupsMapAfterAdd = navLinks.reduce(
          (groupsMap, navLink) => this.addNavLinkToGroup(groupsMap, navGroup, navLink),
          currentGroupsMap
        );

        this.navGroupsMap$.next(navGroupsMapAfterAdd);
      },
      getNavGroupEnabled: () => this.navGroupEnabled,
      registerNavGroupUpdater: (updater$) => {
        this.navGroupUpdaters$$.next([...this.navGroupUpdaters$$.getValue(), updater$]);
        return () => {
          this.navGroupUpdaters$$.next(
            this.navGroupUpdaters$$.getValue().filter((item) => item !== updater$)
          );
        };
      },
    };
  }
  async start({
    navLinks,
  }: {
    navLinks: ChromeNavLinks;
  }): Promise<ChromeNavGroupServiceStartContract> {
    this.navLinks$ = navLinks.getNavLinks$();
    return {
      getNavGroupsMap$: () => this.getSortedNavGroupsMap$(),
      getNavGroupEnabled: () => this.navGroupEnabled,
    };
  }
  async stop() {
    this.stop$.next();
    this.navGroupEnabledUiSettingsSubscription?.unsubscribe();
  }
}
