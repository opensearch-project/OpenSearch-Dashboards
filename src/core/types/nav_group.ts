/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiIconType } from '@elastic/eui/src/components/icon/icon';

/**
 * There are two types of navGroup:
 * 1: system nav group, like data administration / settings and setup
 * 2: use case group, like observability.
 *
 * by default the nav group will be regarded as use case group.
 */
export enum NavGroupType {
  SYSTEM = 'system',
}

export enum NavGroupStatus {
  Visible,
  Hidden,
}

/** @public */
export interface ChromeNavGroup {
  id: string;
  title: string;
  description: string;
  order?: number;
  icon?: EuiIconType;
  type?: NavGroupType;

  status?: NavGroupStatus;
}
