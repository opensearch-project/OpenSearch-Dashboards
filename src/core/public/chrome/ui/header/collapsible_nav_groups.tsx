/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './collapsible_nav_group_enabled.scss';
import { EuiFlexItem, EuiSideNavItemType, EuiSideNav, EuiText } from '@elastic/eui';
import { i18n } from '@osd/i18n';
import React from 'react';
import classNames from 'classnames';
import { ChromeNavLink } from '../..';
import { InternalApplicationStart } from '../../../application/types';
import { createEuiListItem } from './nav_link';
import { getOrderedLinksOrCategories, LinkItem, LinkItemType } from '../../utils';

export interface NavGroupsProps {
  navLinks: ChromeNavLink[];
  suffix?: React.ReactElement;
  style?: React.CSSProperties;
  appId?: string;
  navigateToApp: InternalApplicationStart['navigateToApp'];
  onNavItemClick: (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    navItem: ChromeNavLink
  ) => void;
}

const titleForSeeAll = i18n.translate('core.ui.primaryNav.seeAllLabel', {
  defaultMessage: 'See all...',
});

const LEVEL_FOR_ROOT_ITEMS = 1;

export function NavGroups({
  navLinks,
  suffix,
  style,
  appId,
  navigateToApp,
  onNavItemClick,
}: NavGroupsProps) {
  const createNavItem = ({
    link,
    className,
  }: {
    link: ChromeNavLink;
    className?: string;
  }): EuiSideNavItemType<{}> => {
    const euiListItem = createEuiListItem({
      link,
      appId,
      dataTestSubj: `collapsibleNavAppLink-${link.id}`,
      navigateToApp,
      onClick: (event) => {
        onNavItemClick(event, link);
      },
    });

    return {
      id: `${link.id}-${link.title}`,
      name: <EuiText size="xs">{link.title}</EuiText>,
      onClick: euiListItem.onClick,
      href: euiListItem.href,
      emphasize: euiListItem.isActive,
      className: `nav-link-item ${className || ''}`,
      buttonClassName: 'nav-link-item-btn',
      'data-test-subj': euiListItem['data-test-subj'],
      'aria-label': link.title,
    };
  };
  const createSideNavItem = (
    navLink: LinkItem,
    level: number,
    className?: string
  ): EuiSideNavItemType<{}> => {
    if (navLink.itemType === LinkItemType.LINK) {
      if (navLink.link.title === titleForSeeAll) {
        const navItem = createNavItem({
          link: navLink.link,
        });

        return {
          ...navItem,
          name: <EuiText color="success">{navItem.name}</EuiText>,
          emphasize: false,
        };
      }

      return createNavItem({
        link: navLink.link,
        className,
      });
    }

    if (navLink.itemType === LinkItemType.PARENT_LINK && navLink.link) {
      const props = createNavItem({ link: navLink.link });
      const parentItem = {
        ...props,
        forceOpen: true,
        /**
         * The href and onClick should both be undefined to make parent item rendered as accordion.
         */
        href: undefined,
        onClick: undefined,
        className: classNames(props.className, 'nav-link-parent-item'),
        buttonClassName: classNames(props.buttonClassName, 'nav-link-parent-item-button'),
        items: navLink.links.map((subNavLink) =>
          createSideNavItem(subNavLink, level + 1, 'nav-nested-item')
        ),
      };
      /**
       * OuiSideBar will never render items of first level as accordion,
       * in order to display accordion, we need to render a fake parent item.
       */
      if (level === LEVEL_FOR_ROOT_ITEMS) {
        return {
          className: 'nav-link-fake-item',
          buttonClassName: 'nav-link-fake-item-button',
          name: '',
          items: [parentItem],
          id: `fake_${props.id}`,
        };
      }

      return parentItem;
    }

    if (navLink.itemType === LinkItemType.CATEGORY) {
      return {
        id: navLink.category?.id ?? '',
        name: <div className="nav-link-item">{navLink.category?.label ?? ''}</div>,
        items: navLink.links?.map((link) => createSideNavItem(link, level + 1)),
        'aria-label': navLink.category?.label,
      };
    }

    return {} as EuiSideNavItemType<{}>;
  };
  const orderedLinksOrCategories = getOrderedLinksOrCategories(navLinks);
  const sideNavItems = orderedLinksOrCategories
    .map((navLink) => createSideNavItem(navLink, LEVEL_FOR_ROOT_ITEMS))
    .filter((navItem) => !!navItem.id);

  return (
    <EuiFlexItem style={style}>
      <EuiSideNav items={sideNavItems} isOpenOnMobile />
      {suffix}
    </EuiFlexItem>
  );
}
