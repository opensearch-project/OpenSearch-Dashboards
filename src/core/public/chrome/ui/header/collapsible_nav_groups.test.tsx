/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { NavGroups } from './collapsible_nav_groups';
import { ChromeRegistrationNavLink } from '../../nav_group';
import { ChromeNavLink } from '../../nav_links';

describe('<NavGroups />', () => {
  const getMockedNavLink = (
    navLink: Partial<ChromeNavLink & ChromeRegistrationNavLink>
  ): ChromeNavLink & ChromeRegistrationNavLink => ({
    baseUrl: '',
    href: '',
    id: '',
    title: '',
    ...navLink,
  });
  it('should render correctly', () => {
    const navigateToApp = jest.fn();
    const onNavItemClick = jest.fn();
    const { container, getByTestId, queryByTestId } = render(
      <NavGroups
        navLinks={[
          getMockedNavLink({
            id: 'pure',
            title: 'pure link',
          }),
          getMockedNavLink({
            id: 'subLink',
            title: 'subLink',
            parentNavLinkId: 'pure',
          }),
          getMockedNavLink({
            id: 'link-in-category',
            title: 'link-in-category',
            category: {
              id: 'category-1',
              label: 'category-1',
            },
          }),
          getMockedNavLink({
            id: 'link-in-category-2',
            title: 'link-in-category-2',
            category: {
              id: 'category-1',
              label: 'category-1',
            },
          }),
          getMockedNavLink({
            id: 'sub-link-in-category',
            title: 'sub-link-in-category',
            parentNavLinkId: 'link-in-category',
            category: {
              id: 'category-1',
              label: 'category-1',
            },
          }),
        ]}
        navigateToApp={navigateToApp}
        onNavItemClick={onNavItemClick}
      />
    );
    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('.nav-link-item-btn').length).toEqual(5);
    fireEvent.click(getByTestId('collapsibleNavAppLink-pure'));
    expect(navigateToApp).toBeCalledTimes(0);
    // The accordion is collapsed
    expect(queryByTestId('collapsibleNavAppLink-subLink')).toBeNull();

    // Expand the accordion
    fireEvent.click(getByTestId('collapsibleNavAppLink-pure'));
    fireEvent.click(getByTestId('collapsibleNavAppLink-subLink'));
    expect(navigateToApp).toBeCalledWith('subLink');
  });
});
