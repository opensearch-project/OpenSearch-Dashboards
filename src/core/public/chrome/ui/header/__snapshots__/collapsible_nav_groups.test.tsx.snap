// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavGroups /> should render correctly 1`] = `
<div>
  <div
    class="euiFlexItem"
  >
    <nav
      class="euiSideNav euiSideNav-isOpenMobile"
    >
      <div
        class="euiSideNav__content euiSideNav__contentMobile-xs euiSideNav__contentMobile-s"
        id="euiSideNavContent_generated-id"
      >
        <div
          class="euiSideNavItem euiSideNavItem--root euiSideNavItem--hasChildItems nav-link-fake-item"
        >
          <div
            class="euiSideNavItemButton nav-link-fake-item-button"
          >
            <span
              class="euiSideNavItemButton__content"
            >
              <span
                class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
              />
            </span>
          </div>
          <div
            class="euiSideNavItem__items"
          >
            <div
              class="euiSideNavItem euiSideNavItem--trunk euiSideNavItem--hasChildItems nav-link-item  nav-link-parent-item"
            >
              <button
                aria-label="pure link"
                class="euiSideNavItemButton euiSideNavItemButton-isOpen nav-link-item-btn nav-link-parent-item-button"
                data-test-subj="collapsibleNavAppLink-pure"
                type="button"
              >
                <span
                  class="euiSideNavItemButton__content"
                >
                  <span
                    class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
                    title="pure link"
                  >
                    <div
                      class="euiText euiText--extraSmall"
                    >
                      pure link
                    </div>
                  </span>
                  <span
                    data-euiicon-type="arrowDown"
                  />
                </span>
              </button>
              <div
                class="euiSideNavItem__items"
              >
                <div
                  class="euiSideNavItem euiSideNavItem--branch nav-link-item nav-nested-item"
                >
                  <button
                    aria-label="subLink"
                    class="euiSideNavItemButton euiSideNavItemButton--isClickable nav-link-item-btn"
                    data-test-subj="collapsibleNavAppLink-subLink"
                    type="button"
                  >
                    <span
                      class="euiSideNavItemButton__content"
                    >
                      <span
                        class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
                        title="subLink"
                      >
                        <div
                          class="euiText euiText--extraSmall"
                        >
                          subLink
                        </div>
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiSideNavItem euiSideNavItem--root euiSideNavItem--hasChildItems"
        >
          <div
            aria-label="category-1"
            class="euiSideNavItemButton"
          >
            <span
              class="euiSideNavItemButton__content"
            >
              <span
                class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
                title="category-1"
              >
                <div
                  class="nav-link-item"
                >
                  category-1
                </div>
              </span>
            </span>
          </div>
          <div
            class="euiSideNavItem__items"
          >
            <div
              class="euiSideNavItem euiSideNavItem--trunk euiSideNavItem--hasChildItems nav-link-item  nav-link-parent-item"
            >
              <button
                aria-label="link-in-category"
                class="euiSideNavItemButton euiSideNavItemButton-isOpen nav-link-item-btn nav-link-parent-item-button"
                data-test-subj="collapsibleNavAppLink-link-in-category"
                type="button"
              >
                <span
                  class="euiSideNavItemButton__content"
                >
                  <span
                    class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
                    title="link-in-category"
                  >
                    <div
                      class="euiText euiText--extraSmall"
                    >
                      link-in-category
                    </div>
                  </span>
                  <span
                    data-euiicon-type="arrowDown"
                  />
                </span>
              </button>
              <div
                class="euiSideNavItem__items"
              >
                <div
                  class="euiSideNavItem euiSideNavItem--branch nav-link-item nav-nested-item"
                >
                  <button
                    aria-label="sub-link-in-category"
                    class="euiSideNavItemButton euiSideNavItemButton--isClickable nav-link-item-btn"
                    data-test-subj="collapsibleNavAppLink-sub-link-in-category"
                    type="button"
                  >
                    <span
                      class="euiSideNavItemButton__content"
                    >
                      <span
                        class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
                        title="sub-link-in-category"
                      >
                        <div
                          class="euiText euiText--extraSmall"
                        >
                          sub-link-in-category
                        </div>
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="euiSideNavItem euiSideNavItem--trunk nav-link-item "
            >
              <button
                aria-label="link-in-category-2"
                class="euiSideNavItemButton euiSideNavItemButton--isClickable nav-link-item-btn"
                data-test-subj="collapsibleNavAppLink-link-in-category-2"
                type="button"
              >
                <span
                  class="euiSideNavItemButton__content"
                >
                  <span
                    class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
                    title="link-in-category-2"
                  >
                    <div
                      class="euiText euiText--extraSmall"
                    >
                      link-in-category-2
                    </div>
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </div>
</div>
`;
